### PR TITLE
Add new battery implementation "Renault Twizy" (LV)

### PR DESCRIPTION
**Pull request by M4GNV5**
### What
This PR implements the CAN bus protocol of Renault Twizy Batteries.
~~Not yet tested with real hardware. Also includes some `TODOs` where I was unsure about the correct way to implement something. A review would be much appreciated.~~
Tested and reviewed

### Why
See: #536 

### How
This is just a straightforward decoder for CAN messages sent by a Renault Twizy BMS. The protocol is [documented by the Twizy community here](https://docs.google.com/spreadsheets/d/1gOrG9rnGR9YuMGakAbl4s97a6irHF6UNFV1TS5Ll7MY/edit?gid=0#gid=0).
For now we do not wake up the battery. For a still built-in battery one can simply turn on the vehicle. For a plain battery (without vehice) this is a TODO.

### CODE REVIEW

File: Software/USER_SETTINGS.h [hunk 1]
```diff
@@ -22,6 +22,7 @@
 //#define PYLON_BATTERY
 //#define RJXZS_BMS
 //#define RENAULT_KANGOO_BATTERY
```
Comment (obbardc) [inline Software/USER_SETTINGS.h:24]: Need to add it to github CI to be built.

File: Software/src/battery/RENAULT-TWIZY.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+int16_t max_value(int16_t *entries, size_t len) {
+  int result = INT16_MIN;
+  for(int i = 0; i < len; i++) {
+    if(entries[i] > result)
+      result = entries[i];
+  }
+
+  return result;
+}
+int16_t min_value(int16_t *entries, size_t len) {
```
Comment (obbardc) [inline Software/src/battery/RENAULT-TWIZY.cpp:30]: linebreak above

File: Software/src/battery/RENAULT-TWIZY.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      // TODO: twizy has two pack voltages, assumingly the minimum and maximum measured.
+      // They usually only differ by 0.1V. We use the lower one here
+      // The other one is in the last 12 bit of the CAN packet
+      
+      // pack voltage is encoded as 16 bit integer in dV
+      voltage_dV = (((int16_t)rx_frame.data.u8[5] << 4) | ((int16_t)rx_frame.data.u8[6] >> 4));
+      break;
+    default:
+      break;
+  }
```
Comment (obbardc) [inline Software/src/battery/RENAULT-TWIZY.cpp:124]: Need to set datalayer to new values & also print to USB serial port.

**Review by obbardc — CHANGES_REQUESTED** [review]
small bits

File: Software/USER_SETTINGS.h (hunk 1 above)
Comment (dalathegreat) [reply Software/USER_SETTINGS.h:24]: If the code is not yet tested on real hardware, best to leave it out of CI until it is stable

Comment (obbardc) [reply Software/USER_SETTINGS.h:24]: But CI is there to make sure the main branch builds ? We could get into a position where there are commits which break the build.

File: Software/src/battery/RENAULT-TWIZY.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+void send_can_battery() {
+  // we do not need to send anything to the battery for now
+}
+
+void setup_battery(void) {  // Performs one time setup at startup
+#ifdef DEBUG_VIA_USB
+  Serial.println("Renault Twizy battery selected");
+#endif
+
```
Comment (dalathegreat) [inline Software/src/battery/RENAULT-TWIZY.cpp:135]: Suggestion to add amount of cells to setup_battery, example from 96S LEAF;

`datalayer.battery.info.number_of_cells = 96;`

File: Software/src/battery/RENAULT-TWIZY.cpp (hunk 3 above)
Comment (dalathegreat) [reply Software/src/battery/RENAULT-TWIZY.cpp:124]: Datalayer seems to be set already. Skip printing to USB serial port, we will remove that in the future and promote the "More battery info" page

File: Software/src/battery/RENAULT-TWIZY.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  // TODO: the twizy provides two values: one for the maximum charge provided by the on-board charger
+  //    and one for the maximum charge during recuperation.
+  //    For now we use the lower of the two (usually the charger one)
+  datalayer.battery.status.max_charge_power_W = max_charge_power < max_recup_power ? max_charge_power : max_recup_power;
+
+  datalayer.battery.status.max_discharge_power_W = max_discharge_power;
+
+  datalayer.battery.status.cell_min_voltage_mV = min_value(cellvoltages_mV, sizeof(cellvoltages_mV) / sizeof(*cellvoltages_mV));
+  datalayer.battery.status.cell_max_voltage_mV = max_value(cellvoltages_mV, sizeof(cellvoltages_mV) / sizeof(*cellvoltages_mV));
+
```
Comment (dalathegreat) [inline Software/src/battery/RENAULT-TWIZY.cpp:62]: To get the cellmonitor page working, add the `cellvoltages_mV` array also to the datalayer, example code;

```
          //Map all cell voltages to the global array
          memcpy(datalayer.battery.status.cell_voltages_mV, battery_cell_voltages, 96 * sizeof(uint16_t));
```

File: Software/src/battery/RENAULT-TWIZY.cpp (hunk 3 above)
Comment (obbardc) [reply Software/src/battery/RENAULT-TWIZY.cpp:124]: Oh, OK !

File: Software/USER_SETTINGS.h (hunk 1 above)
Comment (dalathegreat) [reply Software/USER_SETTINGS.h:24]: Not possible to break the normal builds. But yes, it is possible to break the Twizy, before the Twizy is working :)

Comment (M4GNV5) [reply Software/USER_SETTINGS.h:24]: I was able to compile it successfully using the `arduino-cli`. I have not yet received my inverter, but I will make some tests wether the values are read correctly from the Twizy CAN bus.

File: Software/USER_SETTINGS.h [hunk 6]
```diff
@@ -22,6 +22,7 @@
...
+#define RENAULT_TWIZY_BATTERY
```
Comment (dalathegreat) [inline Software/USER_SETTINGS.h:25]: Before merging, uncomment the RENAULT_TWIZY_BATTERY from the USER_SETTINGS.h file, otherwise builds wont pass

**Review by dalathegreat — COMMENTED** [review]
LGTM! Nice starting point for Twizy!

**Review by dalathegreat — APPROVED** [review]

**Comment by M4GNV5** [comment]
I made some successfull tests of the full setup utilizing the Twizy battery protocol and the pylontech LV inverter side.
So far I do not have a real inverter (waiting for delivery), but I managed to dump the produced CAN messages using an USB-CAN adapter.
Below some pictures. The bugs I found during testing have already been commited & approved above. From my side this PR is ready for merging

photo_2024-10-15_16-25-41

Screenshot from 2024-10-15 16-06-01

Screenshot from 2024-10-15 14-39-02

photo_2024-10-15_16-25-42

**Comment by dalathegreat** [comment]
Amazing work @M4GNV5 :raised_hands: 

Lets merge this, and continue in smaller PRs if needed! You can join the Discord via this link: https://discord.gg/amAu64Mu

**Comment by M4GNV5** [comment]
Just in case anyone ever finds this PR wanting to use a Twizy Battery for battery storage.
According to [this schematics](https://github.com/dexterbg/Twizy-Virtual-BMS/blob/master/extras/Twizy-BMS-wiring-scheme.pdf) the connector next to the sevcon HV connector has several pins directly controling the relais for delivering the full battery voltage to three different outputs:

image

**Pull request merged**